### ISSUE TITLE
[Feature] Custom Reports

### ISSUE THREAD

**Issue by shahzadali84** [actualbudget/actual#730]
### Verified feature request does not already exist?

- [X] I have searched and found no existing issue

### 💻

- [X] Would you like to implement this feature?

### Pitch: what problem are you trying to solve?

Hello,

I have just started using Actual. However I noticed that the reports are way too simple and therefore I would like to add some custom reports and use it for my personal use.

If I find it useful I can also add it to the community for wider audience.

I have actual-server checked out locally but I wonder where is the Actual code and how actual-server is getting Actual app? My question is that if I add some custom report, how can I use it just locally for testing and then deploy it for wider audience?

### Describe your ideal solution to this problem

Add reports similar to YNAB

### Teaching and learning

n/a

**Comment by Kidglove57** [comment]
Thank you - I'm just a user not a developer but my big wish would be for an income and expenses account!

**Comment by j-f1** [comment]
Hi! We’d love to have better reports in the app.

> I have actual-server checked out locally but I wonder where is the Actual code and how actual-server is getting Actual app? My question is that if I add some custom report, how can I use it just locally for testing and then deploy it for wider audience?

The code for the frontend (the Actual app you see) is in this repo, and the syncing server is in the actual-server repo. Once you run `yarn install` and `yarn start:browser` in this repo your browser should open and you should be able to make changes and refresh to see the results. You can also connect that instance of Actual to your local server if you want to test with your real data (although I recommend loading from a backup instead so that you don’t get your budget messed up by a bug).

**Comment by rich-howell** [comment]
Hey,

We have an issue #507 here with some report ideas in. 

Cheers
Rich

**Comment by shahzadali84** [comment]
> 

Are there steps to setup the Actual repo locally? I cloned the repo but getting tons of errors when try to do "yarn install" and "yarn start:browser"

**Comment by j-f1** [comment]
Which errors are you getting?

**Comment by shahzadali84** [comment]
> Which errors are you getting?

File attached with errors copied
[errors.txt](https://github.com/actualbudget/actual/files/10915374/errors.txt)

**Comment by carkom** [comment]
FYI, I've done a lot of what you're looking for. Just submitted PR #731

**Comment by j-f1** [comment]
> File attached with errors copied

Thanks! Most of that looks ok, the only one that sticks out to me is this one:


```
➤ YN0009: │ actual@workspace:. couldn't be built successfully (exit code 127, logs can be found here: C:\Users\Shahzad\AppData\Local\Temp\xfs-6f34b1b0\build.log)
```

The file at `C:\Users\Shahzad\AppData\Local\Temp\xfs-6f34b1b0\build.log` should have more details — `patch-package` is probably failing but I’m not sure why that would be happening.

**Comment by shahzadali84** [comment]
> > File attached with errors copied
> 
> Thanks! Most of that looks ok, the only one that sticks out to me is this one:
> 
> ```
> ➤ YN0009: │ actual@workspace:. couldn't be built successfully (exit code 127, logs can be found here: C:\Users\Shahzad\AppData\Local\Temp\xfs-6f34b1b0\build.log)
> ```
> 
> The file at `C:\Users\Shahzad\AppData\Local\Temp\xfs-6f34b1b0\build.log` should have more details — `patch-package` is probably failing but I’m not sure why that would be happening.

The following is from the logs that got created.

**# This file contains the result of Yarn building a package (actual@workspace:.)
# Script name: postinstall

'rm' is not recognized as an internal or external command,
operable program or batch file.
command not found: rm**

**Comment by j-f1** [comment]
Thanks! That helps a lot — opened #733 which should fix that issue once merged.

**Comment by shahzadali84** [comment]
> Thanks! That helps a lot — opened #733 which should fix that issue once merged.

Thanks, I was able to complete the yarn install step. However I noticed errors while executing yarn start:browser command

**E:\actual-server\actual>yarn start:browser**
'sh' is not recognized as an internal or external command,
operable program or batch file.
'bash' is not recognized as an internal or external command,
operable program or batch file.
ERROR: "start:browser-frontend" exited with 1.

**E:\actual-server\actual>yarn start:browser-frontend**
'sh' is not recognized as an internal or external command,
operable program or batch file.

**Comment by j-f1** [comment]
Hm ok, would you be able to try again using WSL? It seems like we don’t really support developing directly on Windows at this point (although that is definitely something we should work towards!)

**Comment by shahzadali84** [comment]
> Hm ok, would you be able to try again using WSL? It seems like we don’t really support developing directly on Windows at this point (although that is definitely something we should work towards!)

Tried with WSL and following is the output. Not sure if I missed anything.

**/mnt/e/actual-server/actual$ sudo yarn start:browser**
events.js:291
      throw er; // Unhandled 'error' event
      ^

Error: spawn ./bin/watch-browser EACCES
    at Process.ChildProcess._handle.onexit (internal/child_process.js:268:19)
    at onErrorNT (internal/child_process.js:470:16)
    at processTicksAndRejections (internal/process/task_queues.js:84:21)
Emitted 'error' event on ChildProcess instance at:
    at Process.ChildProcess._handle.onexit (internal/child_process.js:274:12)
    at onErrorNT (internal/child_process.js:470:16)
    at processTicksAndRejections (internal/process/task_queues.js:84:21) {
  errno: 'EACCES',
  code: 'EACCES',
  syscall: 'spawn ./bin/watch-browser',
  path: './bin/watch-browser',
  spawnargs: []
}
ERROR: "start:browser-frontend" exited with 1.

**Comment by j-f1** [comment]
You may have to do `chmod +x packages/desktop-client/bin/watch-browser`. My guess is that when you first cloned the repo the execute permissions weren’t set up properly by Git.

**Comment by shall0pass** [comment]
I may be ignorant here, but does building in Wsl work the same on a mounted drive rather than building inside /home/user ?  I always have the git directories inside my home folder and never have troubles.

**Comment by shahzadali84** [comment]
> You may have to do `chmod +x packages/desktop-client/bin/watch-browser`. My guess is that when you first cloned the repo the execute permissions weren’t set up properly by Git.

will give it a try thanks

**Comment by shahzadali84** [comment]
> I may be ignorant here, but does building in Wsl work the same on a mounted drive rather than building inside /home/user ? I always have the git directories inside my home folder and never have troubles.

will give it a try thanks

**Comment by youngcw** [comment]
Being worked on in #1791

**Comment by youngcw** [comment]
Custom Reports are nearly complete.  If you have any new comments use #1918

**Comment by MatissJanis** [comment]
Custom reports have been officially released in v24.7.0. Thanks to everyone that contributed!
